core/src/locktower.rs

```diff
@@ -0,0 +1,580 @@
+        let bank_epoch = bank.get_epoch_and_slot_index(bank.slot()).0;
+        if bank_epoch != self.epoch.slot {
+            assert!(bank_epoch > self.epoch.slot, "epoch cannot move backwards");
+            self.epoch = Epoch::new_from_bank(bank);
+        }
+    }
+
+    pub fn record_vote(&mut self, slot: u64) {
+        self.lockouts.process_vote(Vote { slot });
+    }
+
+    pub fn calculate_weight(&self, stake_lockouts: &HashMap<u64, StakeLockout>) -> u128 {
+        let mut sum = 0u128;
+        let root_slot = self.lockouts.root_slot.unwrap_or(0);
+        for (slot, stake_lockout) in stake_lockouts {
+            if self.lockouts.root_slot.is_some() && *slot <= root_slot {
+                continue;
+            }
+            sum += u128::from(stake_lockout.lockout) * u128::from(stake_lockout.stake)
+        }
+        sum
+    }
+
+    pub fn check_already_voted(&self, slot: u64) -> bool {
+        for vote in &self.lockouts.votes {
+            if vote.slot == slot {
+                return true;
+            }
+        }
+        false
+    }
+
+    pub fn check_vote_lockout(
+        &self,
+        slot: u64,
+        flat_children: &HashMap<u64, HashSet<u64>>,
+    ) -> bool {
+        let mut lockouts = self.lockouts.clone();
+        lockouts.process_vote(Vote { slot });
+        for vote in &lockouts.votes {
+            if vote.slot == slot {
+                continue;
+            }
+            if !flat_children[&vote.slot].contains(&slot) {
+                return false;
+            }
+        }
+        if let Some(root) = lockouts.root_slot {
+            flat_children[&root].contains(&slot)
+        } else {
+            true
+        }
+    }
+
+    pub fn check_vote_stake_threshold(
+        &self,
+        slot: u64,
+        stake_lockouts: &HashMap<u64, StakeLockout>,
+    ) -> bool {
+        let mut lockouts = self.lockouts.clone();
+        lockouts.process_vote(Vote { slot });
+        let vote = lockouts.nth_recent_vote(self.threshold_depth);
+        if let Some(vote) = vote {
+            if let Some(fork_stake) = stake_lockouts.get(&vote.slot) {
+                (fork_stake.stake as f64 / self.epoch.total_staked as f64) > self.threshold_size
+            } else {
+                false
+            }
+        } else {
+            true
+        }
+    }
+
+    //lockout is accumulated for all the parent forks
+    fn insert_fork_tree_lockouts(
+        stake_lockouts: &mut HashMap<u64, StakeLockout>,
+        vote: &Lockout,
+        flat_parents: &HashMap<u64, HashSet<u64>>,
+    ) {
+        let mut fork_tree = vec![vote.slot];
+        fork_tree.extend(&flat_parents[&vote.slot]);
+        for slot in fork_tree {
+            let entry = &mut stake_lockouts.entry(slot).or_insert(StakeLockout {
+                lockout: 0,
+                stake: 0,
+            });
+            entry.lockout += vote.lockout();
+        }
+    }
+    //stake size is the same for all the parent forks
```

Comment:
Move this comment inside the function or add it as a "Note: " before a function description.